File: desktop/frontend/src/plugins/builtin/settings/providers/application/modelIdentity.ts
import {
  MAXIMUM_MODEL_IDENTITY_CHARACTERS,
  MAXIMUM_PROVIDER_IDENTITY_CHARACTERS,
  MAXIMUM_REASONING_EFFORT_IDENTITY_CHARACTERS,
} from "@flame/runtime-contract/wire";

const NON_IDENTITY_CHARACTER = /[\p{C}\p{Z}]/u;
type IdentityKind = "provider" | "model" | "reasoningEffort";

function validateIdentity(kind: IdentityKind, value: string, maximumCharacters: number): void {
  if (value.length === 0) throw new Error(`${kind}_identity_empty`);
  if (Array.from(value).length > maximumCharacters) {
    throw new Error(`${kind}_identity_exceeds_${maximumCharacters}_characters`);
  }
  if (NON_IDENTITY_CHARACTER.test(value)) {
    throw new Error(`${kind}_identity_not_canonical`);
  }
}

export function validateProviderIdentity(value: string): void {
  validateIdentity("provider", value, MAXIMUM_PROVIDER_IDENTITY_CHARACTERS);
}

export function validateModelIdentity(value: string): void {
  validateIdentity("model", value, MAXIMUM_MODEL_IDENTITY_CHARACTERS);
}

export function validateReasoningEffortIdentity(value: string): void {
  validateIdentity("reasoningEffort", value, MAXIMUM_REASONING_EFFORT_IDENTITY_CHARACTERS);
}

File: runtime/protocol/model_identity.go
package protocol

import (
	runtimeidentity "github.com/Tangerg/flame/runtime/internal/identity"
)

// Public model-identity envelopes used by generated clients and local
// consumers before a value crosses the Runtime boundary.
const (
	MaximumProviderIdentityCharacters        = runtimeidentity.MaximumProviderCharacters
	MaximumModelIdentityCharacters           = runtimeidentity.MaximumModelCharacters
	MaximumReasoningEffortIdentityCharacters = runtimeidentity.MaximumReasoningEffortCharacters
)

// ErrIncompleteModelSelection reports a provider/model pair where only one
// identity was supplied.
var ErrIncompleteModelSelection = runtimeidentity.ErrIncompleteModelSelection

// ValidateProviderIdentity reports whether value is an exact provider identity.
func ValidateProviderIdentity(value string) error {
	return runtimeidentity.ValidateProviderIdentity(value)
}

// ValidateModelIdentity reports whether value is an exact model identity.
func ValidateModelIdentity(value string) error {
	return runtimeidentity.ValidateModelIdentity(value)
}

// ValidateReasoningEffortIdentity reports whether value is an exact model-owned
// reasoning-effort identity.
func ValidateReasoningEffortIdentity(value string) error {
	return runtimeidentity.ValidateReasoningEffortIdentity(value)
}

// ValidateModelSelection reports whether provider, model, and reasoning effort
// form an unset or exact complete Runtime model selection.
func ValidateModelSelection(provider, model, reasoningEffort string) error {
	return runtimeidentity.ValidateModelSelection(provider, model, reasoningEffort)
}
